libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -602,44 +649,55 @@ bool IotMqtt_IsSubscribed( IotMqttConnection_t mqttConnection,
         .topicNameLength = topicFilterLength,
         .exactMatchOnly  = true
     };
+    int8_t contextIndex = -1;
+    int8_t matchedIndex = -1;
+
+    /* Getting MQTT Context for the specified MQTT Connection. */
+    contextIndex = _IotMqtt_getContextIndexFromConnection( mqttConnection );
 
     /* Prevent any other thread from modifying the subscription list while this
      * function is running. */
-    IotMutex_Lock( &( mqttConnection->subscriptionMutex ) );
-
-    /* Search for a matching subscription. */
-    pSubscriptionLink = IotListDouble_FindFirstMatch( &( mqttConnection->subscriptionList ),
-                                                      NULL,
-                                                      _topicMatch,
-                                                      &topicMatchParams );
-
-    /* Check if a matching subscription was found. */
-    if( pSubscriptionLink != NULL )
+    if( IotMutex_Take( &( connToContext[ contextIndex ].subscriptionMutex ) ) == true )
     {
-        pSubscription = IotLink_Container( _mqttSubscription_t, pSubscriptionLink, link );
+        matchedIndex = IotMqtt_FindFirstMatch( &( connToContext[ contextIndex ].subscriptionArray ), 0
+                                               , &topicMatchParams );
 
-        /* Copy the matching subscription to the output parameter. */
-        if( pCurrentSubscription != NULL )
+        /* Check if a matching subscription was found. */
+        if( matchedIndex != -1 )
         {
-            pCurrentSubscription->pTopicFilter = pTopicFilter;
-            pCurrentSubscription->topicFilterLength = topicFilterLength;
-            pCurrentSubscription->qos = IOT_MQTT_QOS_0;
-            pCurrentSubscription->callback = pSubscription->callback;
+            pSubscription = &( connToContext[ contextIndex ].subscriptionArray );
+
+            /* Copy the matching subscription to the output parameter. */
+            if( pCurrentSubscription != NULL )
+            {
+                pCurrentSubscription->pTopicFilter = pTopicFilter;
+                pCurrentSubscription->topicFilterLength = topicFilterLength;
+                pCurrentSubscription->qos = IOT_MQTT_QOS_0;
+                pCurrentSubscription->callback = pSubscription->callback;
+            }
+            else
+            {
+                EMPTY_ELSE_MARKER;
+            }
+
+            status = true;
         }
         else
         {
             EMPTY_ELSE_MARKER;
         }
 
-        status = true;
+        if( IotMutex_Give( &( connToContext[ contextIndex ].subscriptionMutex ) ) == false )
+        {
+            /* Fail to give subscription mutex as no space is available on queue. */
+            status = false;
```

Comment:
Please look for all the places where this can be made one statement.

Reply:
Addressed